lib/trino-parquet/src/main/java/io/trino/parquet/reader/MetadataReader.java

```diff
@@ -126,32 +184,61 @@ public static ParquetMetadata readFooter(ParquetDataSource dataSource)
-                            metaData.data_page_offset,
-                            metaData.dictionary_page_offset,
-                            metaData.num_values,
-                            metaData.total_compressed_size,
-                            metaData.total_uncompressed_size);
-                    column.setColumnIndexReference(toColumnIndexReference(columnChunk));
-                    column.setOffsetIndexReference(toOffsetIndexReference(columnChunk));
-                    blockMetaData.addColumn(column);
+
+                    ColumnCryptoMetaData cryptoMetaData = columnChunk.getCrypto_metadata();
+                    ColumnPath columnPath = null;
+                    boolean encryptedMetadata = false;
+
+                    if (null == cryptoMetaData) { // Plaintext column
+                        columnPath = getPath(metaData);
+                        if (null != fileDecryptor && !fileDecryptor.plaintextFile()) {
+                            // mark this column as plaintext in encrypted file decryptor
+                            fileDecryptor.setColumnCryptoMetadata(columnPath, false, false, (byte[]) null, columnOrdinal);
+                        }
+                    }
+                    else {  // Encrypted column
+                        if (cryptoMetaData.isSetENCRYPTION_WITH_FOOTER_KEY()) { // Column encrypted with footer key
+                            if (!encryptedFooterMode) {
+                                throw new ParquetCryptoRuntimeException("Column encrypted with footer key in file with plaintext footer");
+                            }
+                            if (null == metaData) {
+                                throw new ParquetCryptoRuntimeException("ColumnMetaData not set in Encryption with Footer key");
+                            }
+                            if (null == fileDecryptor) {
+                                throw new ParquetCryptoRuntimeException("Column encrypted with footer key: No keys available");
+                            }
+                            columnPath = getPath(metaData);
+                            fileDecryptor.setColumnCryptoMetadata(columnPath, true, true, (byte[]) null, columnOrdinal);
+                        }
+                        else { // Column encrypted with column key
+                            encryptedMetadata = true;
+                        }
+                    }
+
+                    try {
+                        if (encryptedMetadata) {
+                            // TODO: We decrypted data before filter projection. This could send unnecessary traffic to KMS.
+                            //  In parquet-mr, it uses lazy decyrption but that required to change ColumnChunkMetadata. We will improve it alter.
+                            metaData = decryptMetadata(rowGroup, cryptoMetaData, columnChunk, fileDecryptor, columnOrdinal);
```

Comment:
Why don't we use `ColumnChunkMetaData#getWithEncryptedMetadata` from parquet-mr ?
It looks like using `EncryptedColumnChunkMetaData` to do the decryption when needed would avoid the need for `HiddenColumnChunkMetaData`.